### PR TITLE
Design thoughts for build images

### PR DESCRIPTION
**Pull request by cmeister2**
@xquery: would you be able to take a look at this proposal? Given as you're currently also doing some work in the area of Docker.

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+This document covers the design thoughts behind build images for speeding up Travis CI and a proposal for how to implement things going forward.
+
+## Background
+
+Travis CI currently uses standard Ubuntu containers to build curl. There are lots and lots of entries in the curl build matrix and currently a single job takes a long time to complete.
+
+Part of this is due to downloading and compiling dependencies (e.g. brotli, wolfssl). There's also elapsed time taken in downloading and installing packages for compilation.
+
+Using Docker containers, we can create a build image whereby the dependencies are already downloaded and installed, and compiled where necessary. Theoretically it also means developers can have the same build environment as the CI jobs, which is good for reproducibility.
```
Comment (xquery) [inline DESIGN.md:11]: yes, the docker image being a declarative definition checked into SCM means we can manage/track changes in environment with absolute precision - for me this is docker main use case (eg. in dev) ... production usage is completely different side of the coin.

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+## Design thoughts
+
+- The build images should be based on Ubuntu Trusty, as Travis CI currently use Ubuntu Trusty as their base image.
```
Comment (xquery) [inline DESIGN.md:15]: I think it is reasonable that the first step is to keep things similar (eg. using Ubuntu Trusty) and increment from there.

File: DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  - An alternative could be Centos - any Linux distribution is likely fine. It might be that we choose Trusty to start with then, then move onto Centos for testing different distributions.
+  - Another alternative is Xenial as many of the builds explicitly request the Xenial distribution.
+- Build images will be strictly versioned using semver for reproducibility. The "latest" tag will not be used.
```
Comment (xquery) [inline DESIGN.md:18]: we have debated 'latest' tag at my company - at best it is something to be considered for production/consumption ... here is a useful link calling it out as an antipattern https://vsupalov.com/docker-latest-tag/

File: DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+- clang-7 (dist:xenial) T=fuzzer
+- clang-7 (dist:xenial) T=tidy
+- clang-7 (dist:xenial) T=scan-build
+- clang-7 (dist:xenial, many sanitization options) T=debug
+
+There are several OSX builds which I don't intend to run through Docker.
+
+It probably makes sense to have an individual build image for each of these different options, caching the dependencies of each. Docker builds are "cheap".
+
+It might make sense to have several stages of build:
```
Comment (xquery) [inline DESIGN.md:52]: this makes sense to me

**Comment by xquery** [comment]
generally looks good to me ...  

* Think about how to represent different build configuration options (compiler flags) for easy composition - there are lots of tools (buildkit, etc) ... though I prefer 'home brew' approach as first step
* I wonder if we should consider separating the build and testing of curl eg. via multi stage or interim containers - that way we could fiddle with docker runtime settings to emulate starved resources, etc ... additionally we can emulate degraded network conditions with docker-compose (https://medium.com/tenable-techblog/simulating-enterprise-networks-in-development-using-the-docker-networking-stack-94bf547743c9)

File: DESIGN.md (hunk 2 above)
Comment (cmeister2) [reply DESIGN.md:15]: Yep. I'm kinda inclined to use Xenial given it's the majority image of the build jobs now. @bagder; no idea if you want to weigh in here about which distribution you'd prefer to build on by default?

File: DESIGN.md (hunk 3 above)
Comment (cmeister2) [reply DESIGN.md:18]: Given this is all basically for the CI of curl, I don't believe that we should support :latest for these images. There's no upside. 95% of the time people won't even be using these images (though they can if they want!)

**Comment by cmeister2** [comment]
I think for the moment teasing apart the testing from the rest of curl would be a tricky prospect. Probably not best to address that right now.

WRT build configuration options, we're hopefully just replicating what the Travis builds already do (that's the initial aim anyway). Once we get parity, it's worth working out in what direction the next steps are (if any)

**Comment by cmeister2** [comment]
I've merged this design now, but can still chat here.

So at work we basically have a repository per docker image, and each repository is set up and versioned separately. So in this model we'd have for example

```
base_xenial@v1.0.0 => base_gcc8@v1.0.3 => gcc8_normal@v2.0.4
```

The benefits of this approach are:
- you don't have to rebuild images that don't require fixes (unnecessary rebuilding can fail down the line if the repos no longer exist)
- you can be more precise about versioning.

The downsides:
- you have a repository per image.
- changes in base images may cause a ripple of new builds.

In the github model, individual repos would probably necessitate creating another organization (`curl-ci`?) and having all the new repositories there, which is a bit strange but probably manageable. TBH I could probably just do that anyway and just get buy-in at the last stage (which is actually using them in the main builds).

The alternative is just having a coarse VERSION at the top level of this repository, then regenerating all images and tagging them with that version. Quick and dirty and some images may not even see any changes at all. It will be simpler for the curl CI build system to operate using a single VERSION number across the board.

@xquery - thoughts on any of the above? I think I wanted the first of these, but have talked myself around to the coarse VERSION number, simply because updating a single number in the curl CI builds is very appealing.

**Comment by xquery** [comment]
I prefer a single repo with every image using a single coarse level version. 

As these images are not for wide distribution we could consider embedding timestamp into version number which removes any ambiguity tracing back commit to this repo.

File: DESIGN.md (hunk 2 above)
Comment (bagder) [reply DESIGN.md:15]: I don't have any strong opinion either way. For our travis CI builds we strive to switch to as recent versions as possible when we can.

Comment (cmeister2) [reply DESIGN.md:15]: Ok. With this approach we should theoretically be able to run the newest Ubuntu version if desired. For now, let's do Xenial, but with an eye to easily replacing this in future.

**Comment by cmeister2** [comment]
Sounds good. I will continue on my current plan then, and hope to get something sensible in place shortly for testing.

**Pull request merged**